### DISCUSSION TITLE
Use TypeScript "as const" instead of deepFreeze?

### DISCUSSION THREAD

**Discussion by AjaxSolutions** [pmndrs/jotai#2068] [Q&A] [answered]
@dai-shi ,

Can you use TypeScript "as const" instead of the `deepFreeze` function to assert state as deeply immutable?

https://github.com/pmndrs/jotai/blob/2bf87a6e2564d11aa55671297feb4b055a2ed7a4/src/vanilla/utils/freezeAtom.ts#L8

```
const state = get(anAtom);
return state as const;
```
This change would enforce immutability at the type-level and be simpler and more readable than the current approach.

Do you need runtime immutability enforcement in this case?

Also, see DeepReadonly:

https://github.com/ts-essentials/ts-essentials/tree/master/lib/deep-readonly

**Comment by dai-shi** [comment, marked as answer]
freezeAtom is a runtime enforcement and it doesn't expect TS. I don't think people use it very often.

Yeah, DeepReadonly might help. We could make a new util, but maybe a type util would be enough. Not sure if people use it, though.

Mutating an atom value is technically possible and it sometimes is a hack. We could make an atom value readonly by default, but it would cause troubles and it's not "correct" anyway because the value is not "frozen" in JS.
